### PR TITLE
fix: optimize absorb_and_eliminate and remove_complements

### PR DESCRIPTION
**Pull request by barakalon**
Some of the simplification rules become _very_ expensive on deeply nested conditions (e.g. a1 AND a2 AND ... AND a100)

Here's something that'll probably be controversial - this sets a configurable threshold to not even bother trying to simplify these.

### CODE REVIEW

**Review by georgesittas — APPROVED** [review]
Let's test this?

File: sqlglot/optimizer/simplify.py [hunk 1]
```diff
@@ -407,28 +409,49 @@ def absorb_and_eliminate(expression, root=True):
...
-                        a.replace(aa)
-                        b.replace(aa)
-                    elif ab in rhs and (is_complement(aa, ba) or is_complement(aa, bb)):
-                        a.replace(ab)
-                        b.replace(ab)
+        ops = tuple(expression.flatten())
+
+        complements = set()
+        subops = defaultdict(list)
+        pairs = defaultdict(list)
```
Comment (VaggelisD) [inline sqlglot/optimizer/simplify.py:416]: The previous algorithm was a bit easier to follow, is it possible to leave a comment explaining the usecase of these collections?

File: sqlglot/optimizer/simplify.py [hunk 2]
```diff
@@ -407,28 +409,49 @@ def absorb_and_eliminate(expression, root=True):
...
+
+        # Build the indexes
+        for op in ops:
+            complements.add(op)
+
```
Comment (VaggelisD) [inline sqlglot/optimizer/simplify.py:421]: Can this be rewritten to merge the 2 loops in 1? For instance:
- In the first loop, switch the order of the `if` and `else` e.g.:

```Python3
        for op in ops:
            complements.add(op)
            
            if not isinstance(op, kind):
               subops[op].append({op})
              continue
           ...
```
- The second loop can maybe be moved upwards

File: sqlglot/optimizer/simplify.py [hunk 3]
```diff
@@ -407,28 +409,49 @@ def absorb_and_eliminate(expression, root=True):
...
+                for i in subset:
+                    subops[i].append(subset)
+
+                a, b = op.unnest_operands()
+                if isinstance(a, exp.Not):
+                    pairs[frozenset((a.this, b))].append((op, b))
+                if isinstance(b, exp.Not):
+                    pairs[frozenset((a, b.this))].append((op, a))
+            else:
+                subops[op].append({op})
```
Comment (VaggelisD) [inline sqlglot/optimizer/simplify.py:433]: The second loop seems to go over only over the `isinstance(op, kind)` ops, are the `else` subops being utilized?

**Review by VaggelisD — COMMENTED** [review]
Very interesting! I had a couple of questions:

- I'm presuming that the cost of the simplification on deeply nested conds was exposed by a benchmark, if so, could you share the results & the test case? I've been thinking that we could benefit from amassing a bench suite that stresses each component individually, it will allows to catch regressions and prove the effectiveness of an optimization against historical runs 

- How is the threshold baked in? I may be missing it

**Comment by barakalon** [comment]
@VaggelisD Ah! sorry - this became a WIP after a side discussion on slack. If you look at the original commit, my initial approach had the threshold. Then I decided to try and optimize the existing code.

Planning on doing some cleaning, commenting, benchmarking, etc. today

**Comment by barakalon** [comment]
Added a new benchmarking script for the optimizer.

With this branch:
|       Benchmark |    Duration (s) |
| --------------- | --------------- |
|            tpch |          0.6284 |
|           tpcds |          8.1603 |
|    condition_10 |          0.0042 |
|   condition_100 |          0.0978 |
|  condition_1000 |          4.0751 |

With main:
|       Benchmark |    Duration (s) |
| --------------- | --------------- |
|            tpch |          0.6513 |
|           tpcds |          8.4834 |
|    condition_10 |          0.0048 |
|   condition_100 |          0.2624 |
|   condition_1000 |    ... |

I skipped condition_1000 for main after waiting a few minutes for it to complete.

**Comment by georgesittas** [comment]
Sick

File: sqlglot/optimizer/simplify.py [hunk 4]
```diff
@@ -407,28 +409,49 @@ def absorb_and_eliminate(expression, root=True):
     if isinstance(expression, exp.Connector) and (root or not expression.same_parent):
         kind = exp.Or if isinstance(expression, exp.And) else exp.And
```
[the remark this answers is not in the document]
Comment (barakalon) [reply sqlglot/optimizer/simplify.py:410]: Sure, I can add that

File: sqlglot/optimizer/simplify.py (hunk 2 above)
Comment (barakalon) [reply sqlglot/optimizer/simplify.py:421]: sure

File: sqlglot/optimizer/simplify.py (hunk 1 above)
Comment (barakalon) [reply sqlglot/optimizer/simplify.py]: Added some comments

File: sqlglot/optimizer/simplify.py (hunk 3 above)
Comment (barakalon) [reply sqlglot/optimizer/simplify.py]: Added a comment

File: sqlglot/optimizer/simplify.py (hunk 2 above)
Comment (VaggelisD) [reply sqlglot/optimizer/simplify.py]: Not sure if this is still WIP or if I confused you, but I meant the 2 `for op in ops` loops here

Comment (VaggelisD) [reply sqlglot/optimizer/simplify.py]: I'm guessing it's actually not feasible because you need the indexes built before looping again, nevermind

Comment (barakalon) [reply sqlglot/optimizer/simplify.py]: Ah, yeah. Exactly

**Comment by tobymao** [comment]
@barakalon do you want to also have the maximal ops bypass? 4 seconds is still a long time

**Comment by barakalon** [comment]
Max ops would be great for my use case, yeah. If you're down, I'll add it.

**Comment by tobymao** [comment]
@barakalon yea let's add it

**Review by georgesittas — APPROVED** [review]
great work 🎉

File: sqlglot/optimizer/simplify.py [hunk 5]
```diff
@@ -404,31 +450,63 @@ def absorb_and_eliminate(expression, root=True):
...
-                        a.replace(aa)
-                        b.replace(aa)
-                    elif ab in rhs and (is_complement(aa, ba) or is_complement(aa, bb)):
-                        a.replace(ab)
-                        b.replace(ab)
+        ops = tuple(expression.flatten())
+
+        # Initialize lookup tables:
+        # Set of all operands, used to find complements for absorption.
+        hashed = set()
```
[the remark this answers is not in the document]
Comment (barakalon) [reply sqlglot/optimizer/simplify.py]: named it `op_set` 🤷

**Review by tobymao — APPROVED** [review]

**Pull request merged**